CodeListLibrary_project/clinicalcode/entity_utils/search_utils.py

```diff
@@ -0,0 +1,430 @@
+    template_ids = list(templates.values_list('id', flat=True))
+
+    # Get all entities assoc. with the template(s) requested
+    entities = GenericEntity.objects.filter(template__in=template_ids).all()
+    entities = PublishedGenericEntity.objects.filter(entity__in=entities).order_by('-created').distinct()
+    
+    # Gather request params for the filters across template when not single search
+    template_filters = [ ]
+    if not is_single_search:
+        filters = set([])
+        for items in templates.values_list('entity_filters', flat=True).distinct():
+            filters = filters | set(items)
+        template_filters = list(filters)
+
+        template_fields = list(layouts.values())
+        if len(template_fields) > 0:
+            template_fields = template_utils.get_layout_fields(template_fields[0])
+
+    # Gather metadata filter params
+    metadata_filters = [key for key, value in constants.metadata.items() if 'search' in value and 'filterable' in value.get('search')]
+    
+    # Build query from filters
+    query = { }
+    for param, data in getattr(request, method).items():
+        if param in metadata_filters:
+            if template_utils.is_single_search_only(constants.metadata, param) and not is_single_search:
+                continue
+            apply_param_to_query(query, constants.metadata, param, data, force_term=force_term)
+        elif param in template_filters and not is_single_search:
+            if template_fields is None:
+                continue
+            apply_param_to_query(query, template_fields, param, data, is_dynamic=True, force_term=force_term)
+    
+    # Collect all entities that are (1) published and (2) match request parameters
+    entities = GenericEntity.history.filter(
+        Q(**query) \
+        & \
+        Q(
+            history_id__in=list(entities.values_list('entity_history_id', flat=True)),
+            id__in=list(entities.values_list('entity_id', flat=True))
+        )
+    )
+
+    # Prepare order clause
+    search_order = try_get_param(request, 'order_by', '1', method)
+    should_order_search = search_order == '1'
+    search_order = template_utils.try_get_content(constants.ORDER_BY, search_order)
+    if search_order is None:
+        search_order = constants.ORDER_BY['1']
+    
+    # Apply any search param if present
+    search = try_get_param(request, 'search', None)
+    if search is not None:
+        entities = search_entities(queryset=entities, search_query=search, order_by_relevance=should_order_search, fuzzy=True, min_threshold=0.1)
+
+    # Reorder by user selection
+    if search_order != constants.ORDER_BY['1']:
+        entities = entities.order_by(search_order.get('clause'))
+    else:
+        if search is None:
+            entities = entities.all().extra(
+                select={'true_id': """CAST(REGEXP_REPLACE(id, '[a-zA-Z]+', '') AS INTEGER)"""}
+            ).order_by('true_id', 'id')
```

Comment:
We should really change this - needed due to change to ID from int to char field, but not the best solution.

Might be slightly faster to use the following code, but likely still doesn't scale well:
```sql
ORDER BY SUBSTRING(id, '^[0-9]+')::int, SUBSTRING(id, '[^0-9]*$')
```